### PR TITLE
fix: rework boolean column #315

### PR DESCRIPTION
**Pull request by dotpointer**
This follows the request in #315 to use paper-autocomplete for the column fields instead of the current setup.

Issue is #315.

### CODE REVIEW

File: cosmoz-omnitable-column-boolean.js [hunk 1]
```diff
@@ -83,6 +112,71 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
+							.map(item => {
+								let text,
+									value;
+
+								if (typeof item === 'object') {
+									text = item['text'];
+									value = item['value'];
+								} else {
+									value = text = item.toString();
+								}
```
Comment (cristinecula) [inline cosmoz-omnitable-column-boolean.js]: Isn't `item` always an `object`? Can this callback be simplified?

File: cosmoz-omnitable-column-boolean.js [hunk 2]
```diff
@@ -21,23 +24,44 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
 
 		<template class="edit-cell" strip-whitespace>
-			<paper-dropdown-menu>
-				<paper-listbox class="dropdown-content" slot="dropdown-content"
-					attr-for-selected="value" selected="[[ _computeSelected(item, valuePath) ]]" on-selected-item-changed="_valueChanged">
-					<paper-item value="true">[[ trueLabel ]]</paper-item>
-					<paper-item value="false">[[ falseLabel ]]</paper-item>
-				</paper-listbox>
-			</paper-dropdown-menu>
+			<paper-autocomplete
```
[the remark this answers is not in the document]
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: Added it, worked, thanks.

File: cosmoz-omnitable-column-boolean.js [hunk 3]
```diff
@@ -114,28 +208,67 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
 			return;
 		}
 		this._filterChangedFromAbove = true;
-		this._textFilter = typeof filter === 'boolean' ? filter.toString() : null;
+		if (filter === true) {
+			this._textFilter = this.trueLabel;
+		} else if (filter === false) {
+			this._textFilter = this.falseLabel;
+			this._textFilter = false;
```
[the remark this answers is not in the document]
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: No, removed it.

File: cosmoz-omnitable-column-boolean.js [hunk 4]
```diff
@@ -114,28 +208,67 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
-	_applySingleFilter(item) {
+	/**
+	* Determine if a filter should be enabled or not.
+	* @param {string} filter Filter text.
+	* @param {object} item Column data.
+	* @returns {boolean} Whether the filter should be enabled or not.
+	*/
+	_applySingleFilter(filter, item) {
 		return this.get(this.valuePath, item) === this.filter;
 	}
```
[the remark this answers is not in the document]
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: Yes, used it now.

File: cosmoz-omnitable-column-boolean.js [hunk 5]
```diff
@@ -155,32 +288,77 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
+	* @param {object} item Column data.
+	* @param {string} valuePath Value path in column data.
+	* @returns {string} Column in XLSX value format.
+	*/
 	toXlsxValue(item, valuePath = this.valuePath) {
 		if (!valuePath) {
 			return '';
 		}
-		return this.get(valuePath, item) ? this.trueLabel : this.falseLabel;
+		// old return this.get(valuePath, item) ? this.trueLabel : this.falseLabel;
```
[the remark this answers is not in the document]
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: Dropped.

File: cosmoz-omnitable-column-boolean.js [hunk 6]
```diff
@@ -155,32 +288,77 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
+	* @param {string} title Column title.
+	* @param {string} filter Filter text.
+	* @returns {string} Tooltip text.
+	*/
+	_computeTooltip(title, filter) {
+		if (filter === true) {
+			return this.trueLabel;
+		} else if (filter === false) {
+			return '';
+			return this.falseLabel;
```
[the remark this answers is not in the document]
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: Yes, removed the line above it.

File: cosmoz-omnitable-column-boolean.js [hunk 7]
```diff
@@ -114,28 +208,67 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
+
 	_computeSelected(item, valuePath) {
 		const value = this.get(valuePath || this.valuePath, item);
-		if (value != null) {
-			return value.toString();
+		return value;
+	}
+
+	_computeText(item, valuePath, trueLabel, falseLabel) {
+		const value = this.get(valuePath || this.valuePath, item);
```
[the remark this answers is not in the document]
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: Yes, I rewrote it to the suggestion on this line and other similar lines too.

File: cosmoz-omnitable-column-boolean.js [hunk 8]
```diff
@@ -93,16 +187,16 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
 			this._filterChangedFromAbove = false;
 			return;
 		}
 		this._filterChangedFromInput = true;
-		if (textFilter) {
-			this.set('filter', textFilter === 'true');
+		if (textFilter === 'true' || textFilter === true) {
+			this.set('filter', true);
+		} else if (textFilter === 'false' || textFilter === false) {
+			this.set('filter', false);
```
[the remark this answers is not in the document]
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: Yes it can, `textFilter` appears to be a textual derivative of `filter`.

`filter` is a boolean, `textFilter` is a string.

File: cosmoz-omnitable-column-boolean.js (hunk 1 above)
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: @cristinecula Yes, I have simplified this section, maybe it can be simplified more.

**Comment by dotpointer** [comment]
Not sure how to fix the test error `error:	Uncaught exception: TypeError: chalk.stripColor is not a function`.

**Comment by JockeCK** [comment]
Tests need to be updated/changed/removed. The boolean ones at least should be fixed in this PR.

**Comment by dotpointer** [comment]
> Tests need to be updated/changed/removed. The boolean ones at least should be fixed in this PR.

The error went away after updating components. Updated the tests.

**Comment by dotpointer** [comment]
Merged commits.

**Comment by cristinecula** [comment]
@dotpointer do you get the same test errors when running the tests locally?

**Comment by dotpointer** [comment]
> @dotpointer do you get the same test errors when running the tests locally?

The test errors disappeared with your fix of the tests.

**Comment by dotpointer** [comment]
@JockeCK Can you review again please now that the tests are updated and the errors has went away?

File: cosmoz-omnitable-column-boolean.js [hunk 9]
```diff
@@ -114,28 +205,71 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
 	_computeSelected(item, valuePath) {
 		const value = this.get(valuePath || this.valuePath, item);
-		if (value != null) {
-			return value.toString();
+		if (value === true) {
+			return 'true';
+		} else if (value === false) {
+			return 'false';
+		}
+		return value;
```
Comment (cristinecula) [inline cosmoz-omnitable-column-boolean.js]: `true.toString() === 'true'` so you could stick with `return value.toString()`

File: cosmoz-omnitable-column-boolean.js [hunk 10]
```diff
@@ -114,28 +205,71 @@ class OmnitableColumnBoolean extends columnMixin(PolymerElement) {
...
 		}
 		this._filterChangedFromAbove = true;
-		this._textFilter = typeof filter === 'boolean' ? filter.toString() : null;
+		if (filter === true) {
+			this._textFilter = this.trueLabel;
+		} else if (filter === false) {
+			this._textFilter = this.falseLabel;
+		} else {
+			this._textFilter = null;
+		}
```
Comment (cristinecula) [inline cosmoz-omnitable-column-boolean.js]: I think you should stick with the old version for these lines. The filter should use the `filter` value not the `label`. The label is for humans, the value is for machines. You should also revert the change to the `filterChanged observer converts boolean to string` test.

File: cosmoz-omnitable-column-boolean.js (hunk 9 above)
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: Made a fix to address this in https://github.com/Neovici/cosmoz-omnitable/pull/340.

File: cosmoz-omnitable-column-boolean.js (hunk 10 above)
Comment (dotpointer) [reply cosmoz-omnitable-column-boolean.js]: Made a fix to address this in https://github.com/Neovici/cosmoz-omnitable/pull/340.
